Install tutorial and README update (#43)
arfc/ghastly: README.md

# Ghastly

Ghastly is a python package that externally couples Discrete Element Method
(DEM) physics simulation tools to neutron transport and depletion tools in

> order to model flowing, pebble-bed High-Temperature Gas-Cooled Reactors (HTGRs).

Reviewer: ```suggestion
(DEM) physics simulation tools to neutron transport and depletion tools
to model flowing pebbles in pebble-bed reactors (PBRs).
```

Reviewer: Also: what do you mean by "flowing"? Since I know your project, I know you are getting at pebbles moving in a reactor, but that's not explicit here. Consider changing to "flowing pebbles in a pebble-bed ..."

Ghastly is compatible with [LAMMPS](https://www.lammps.org/) and 
[OpenMC](https://openmc.org/). Ghastly automatically generates input files and

> can read output files from compatible codes.  The examples/ directory contains

Reviewer: ```suggestion
can read output files from compatible codes.  The `examples/` directory contains
```

Ghastly input files for a generic HTGR design.

Each template has some hard-coded settings, some of which are required to

> create output files needed by Ghastly.  Users unfamiliar with compatible tools

Reviewer: So some have hard-coded settings, and a subset of those hard-coded settings are needed? This is confusing to read imo... Have you documented in your templates what can and should not be edited by users?

Author: I have not - perhaps that would be a good addition to the templates directory as a README?  Or I edit the templates to add comments.

Reviewer: Maybe just another issue for down the road?

should read the documentation carefully before making changes to templates.

Currently, Ghastly supports coupling LAMMPS to OpenMC only, but developing 

> and submitting an update to add further code support via PR is welcome.

Reviewer: "but contributions via a Pull Request is welcome to expand capabilities."

> and submitting an update to add further code support via PR is welcome.

Author: ```suggestion
Currently, Ghastly supports coupling LAMMPS to OpenMC only, but contributions via a Pull Request are welcome to expand capabilities.
```

Note that the paraview-plot.py file is meant to be loaded into your Paraview
GUI python script editor.

## Ghastly can:

> - Fill both upwards and downwards flowing PBRs with an initial bed 

Reviewer: nitpicky, but I don't see anywhere above where you explicitly write out Pebble Bed Reactors before using the acronym... If you think it is fine, I won't argue.

Author: fixed with above change

configuration
- Recirculate downwards flowing (HTGR) pebble beds.

## Ghastly cannot:
- Model PBRs with a central reflector
- Model PBRs with multiple inlet and outlet chutes

# Installation:

> Ghastly is not packaged with any source code for compatible tools.  It is 

Reviewer: > Ghastly is not packaged with any source code for compatible tools.

What does this mean? Can you rewrite in a way that gets the same meaning without potentially inviting more questions?

Author: This is me trying to make it very clear that Ghastly does not come bundled with LAMMPS/OpenMC/any other tool's source code.  This is related to licensing rules and guidelines - some of the licensing rules are more complex than others, and I wanted to help make it clearer (for those that know licensing, at least) where Ghastly falls.

Reviewer: Would this work then?

"Ghastly is not packaged with any additional software for compatible tools."

If so, then it was the use of the word "code" that threw me off, because for some reason I immediately jumped to thinking about source files.

Author: Except I do mean the source code - Ghastly doesn't touch it, isn't packaged with it.  It's a distinct language I'm using for licensing reasons.

recommended, but optional, that users install Ghastly and the DEM and Monte Carlo
tools of choice inside a conda environment.

Activate the conda environment if you choose to use one before installing 
any dependencies. (See the Notes section below for tips about using conda
environments while installing from source with cmake)  The setup-example script
is only meant to serve as an example, and was not written with the intention
of running on an arbitrary system.

## Dependencies and Compatible Tool installation:

### OpenMC:
The quick install guide for OpenMC is [here](https://docs.openmc.org/en/stable/quickinstall.html).
Follow the instructions to build from source.  The first step is to install HDF5.
The setup script included in this level of the repository is an example of how
to install mpi4py, parallel h5py, and parallel HDF5 inside a conda environment.

If you want to use OpenMC with OpenMP or MPI, set 
`-DOpenMC_USE_OPENMP=on` and `-DOPENMC_USE_MPI=on`.  If you are installing 
inside a conda environment, also set `-DCMAKE_INSTALL_PREFIX=${CONDA_PREFIX}` 
and call `HDF5_ROOT=${CONDA_PREFIX}` before calling cmake.

### LAMMPS:

> The documentation page for building LAMMPS begins [here](https://docs.lammps.org/Build.html) here.

Reviewer: ```suggestion
The documentation page for building LAMMPS begins [here](https://docs.lammps.org/Build.html).
```

LAMMPS has many optional packages and build options which will not be discussed
in depth here beyond what should be installed if working with Ghastly.  Users 

> should set `-D BUILD_SHARED_LIBS=yes`, `-D BUILD_MPI=yes`, `-D BUILD_OMP=yes`.  

Reviewer: It's not super clear why these flags need to be set... Can you briefly elaborate?

Author: I am not certain I agree about extra context actually being helpful here - and I think in the case of BUILD_MPI and BUILD_OMP especially, this README tutorial is not the place to explain MPI and OpenMP.

Reviewer: Ah, so the message is: "When building lamps, be sure to set these additional flags so that Ghastly works properly."

If this is true, then feel free to resolve this.

The setup script example also turns the JPEG, PNG, and FFMPEG options on.  
These are not needed for Ghastly to work, though users may prefer to have the option.
As always, setting `-D CMAKE_INSTALL_PREFIX` and `-D CMAKE_PREFIX_PATH` to

> `${CONDA_PREFIX}` is required if using conda environments.  In order to use

Reviewer: This would be a bigger change, but I would actually recommend creating a whole set of instructions (numbered 1., 2., etc.) for installing with and without Conda. Having to read through the entire documentation and filter what I do and don't would be frustrating if I was a user. Just something to consider...

Author: I'm not sure it would need two full bulleted lists, because the installation method without conda can be described as 'whatever the source documentation says' or 'identical to the conda env install, but don't use any of the flags set to some variation of ${CONDA_PREFIX}'.  However, I could make a comment to that effect at the beginning of the install tutorial, and then simplify the language in the rest of the tutorial.

Reviewer: I think that would be good. If I were an undergraduate given a project by my advisor (with very little prior experience), having a straightforward set of instructions for Conda install would be good. Totally on board with telling users they are on their own if they want to build from source.

> `${CONDA_PREFIX}` is required if using conda environments.  In order to use

Reviewer: ```suggestion
These are not needed for Ghastly to work, though users may prefer to have the option.
As always, setting `-D CMAKE_INSTALL_PREFIX` and `-D CMAKE_PREFIX_PATH` to
`${CONDA_PREFIX}` is required if using conda environments.  To use
```

LAMMPS to model pebble flow, users must install the GRANULAR package with 
`PKG_GRANULAR=yes`.  Users should also include the OPENMP package with 
`PKG_OPENMP=yes`.

## Ghastly:

Ghastly can be installed from PyPi for general use.  If you plan to contribute
to Ghastly, you should instead create a fork of the arfc/ghastly repository, 
then clone and install Ghastly in editable mode:

```
git clone git@github.com:YourUsername/ghastly.git
cd ghastly/
python -m pip install --editable .
```

## Notes:

### Installing inside conda environments:
Installing inside a conda environment makes uninstalling Ghastly and its
compatible tools easier, which is useful if anything goes wrong (especially
if you would like to contribute to Ghastly).  Installing packages available
with conda-forge is simple.  For installing software from source into a conda
environment, remember:

> - when a conda environment is active, the environment variable ${CONDA_PREFIX}

Reviewer: ```suggestion
- When a conda environment is active, the environment variable ${CONDA_PREFIX}
```
